File: Graph-Algorithms/GraphAlgorithms.java
/*
 * 🌐 Graph Algorithms - Java Implementation
 * Author: AI Assistant
 * 
 * Description:
 * This Java program contains essential graph algorithms frequently asked
 * in technical interviews at top companies like FAANG, Google, Microsoft, and Amazon.
 * 
 * Each algorithm includes:
 *  - Problem definition
 *  - Clean Java implementation
 *  - Example usage
 *  - Time and Space Complexity
 */

import java.util.*;

public class GraphAlgorithms {
    
    // Graph representation using adjacency list
    static class Graph {
        private int V;
        private List<List<Integer>> adj;
        
        Graph(int V) {
            this.V = V;
            adj = new ArrayList<>();
            for (int i = 0; i < V; i++) {
                adj.add(new ArrayList<>());
            }
        }
        
        void addEdge(int u, int v) {
            adj.get(u).add(v);
        }
        
        void addUndirectedEdge(int u, int v) {
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        
        List<Integer> getNeighbors(int v) {
            return adj.get(v);
        }
        
        int getVertices() { return V; }
    }
    
    // 1️⃣ Breadth-First Search (BFS)
    public static void BFS(Graph graph, int start) {
        boolean[] visited = new boolean[graph.getVertices()];
        Queue<Integer> queue = new LinkedList<>();
        
        visited[start] = true;
        queue.offer(start);
        
        System.out.print("BFS traversal: ");
        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            System.out.print(vertex + " ");
            
            for (int neighbor : graph.getNeighbors(vertex)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.offer(neighbor);
                }
            }
        }
        System.out.println();
    }
    
    // 2️⃣ Depth-First Search (DFS) - Recursive
    public static void DFS(Graph graph, int start) {
        boolean[] visited = new boolean[graph.getVertices()];
        System.out.print("DFS traversal: ");
        DFSRecursive(graph, start, visited);
        System.out.println();
    }
    
    private static void DFSRecursive(Graph graph, int vertex, boolean[] visited) {
        visited[vertex] = true;
        System.out.print(vertex + " ");
        
        for (int neighbor : graph.getNeighbors(vertex)) {
            if (!visited[neighbor]) {
                DFSRecursive(graph, neighbor, visited);
            }
        }
    }
    
    // 3️⃣ DFS - Iterative
    public static void DFSIterative(Graph graph, int start) {
        boolean[] visited = new boolean[graph.getVertices()];
        Stack<Integer> stack = new Stack<>();
        
        stack.push(start);
        System.out.print("DFS (Iterative): ");
        
        while (!stack.isEmpty()) {
            int vertex = stack.pop();
            if (!visited[vertex]) {
                visited[vertex] = true;
                System.out.print(vertex + " ");
                
                for (int neighbor : graph.getNeighbors(vertex)) {
                    if (!visited[neighbor]) {
                        stack.push(neighbor);
                    }
                }
            }
        }
        System.out.println();
    }
    
    // 4️⃣ Detect Cycle in Directed Graph (DFS with colors)
    public static boolean hasCycleDirected(Graph graph) {
        int[] color = new int[graph.getVertices()]; // 0: white, 1: gray, 2: black
        
        for (int i = 0; i < graph.getVertices(); i++) {
            if (color[i] == 0 && hasCycleDFS(graph, i, color)) {
                return true;
            }
        }
        return false;
    }
    
    private static boolean hasCycleDFS(Graph graph, int vertex, int[] color) {
        color[vertex] = 1; // gray
        
        for (int neighbor : graph.getNeighbors(vertex)) {
            if (color[neighbor] == 1) return true; // back edge
            if (color[neighbor] == 0 && hasCycleDFS(graph, neighbor, color)) {
                return true;
            }
        }
        
        color[vertex] = 2; // black
        return false;
    }
    
    // 5️⃣ Detect Cycle in Undirected Graph
    public static boolean hasCycleUndirected(Graph graph) {
        boolean[] visited = new boolean[graph.getVertices()];
        
        for (int i = 0; i < graph.getVertices(); i++) {
            if (!visited[i] && hasCycleUndirectedDFS(graph, i, -1, visited)) {
                return true;
            }
        }
        return false;
    }
    
    private static boolean hasCycleUndirectedDFS(Graph graph, int vertex, int parent, boolean[] visited) {
        visited[vertex] = true;
        
        for (int neighbor : graph.getNeighbors(vertex)) {
            if (!visited[neighbor]) {
                if (hasCycleUndirectedDFS(graph, neighbor, vertex, visited)) {
                    return true;
                }
            } else if (neighbor != parent) {
                return true; // back edge
            }
        }
        return false;
    }
    
    // 6️⃣ Topological Sorting (Kahn's Algorithm)
    public static List<Integer> topologicalSort(Graph graph) {
        int[] inDegree = new int[graph.getVertices()];
        
        // Calculate in-degrees
        for (int i = 0; i < graph.getVertices(); i++) {
            for (int neighbor : graph.getNeighbors(i)) {
                inDegree[neighbor]++;
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < graph.getVertices(); i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }
        
        List<Integer> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            result.add(vertex);
            
            for (int neighbor : graph.getNeighbors(vertex)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        
        return result.size() == graph.getVertices() ? result : new ArrayList<>();
    }
    
    // 7️⃣ Shortest Path - BFS (unweighted graph)
    public static int[] shortestPathBFS(Graph graph, int start) {
        int[] distance = new int[graph.getVertices()];
        Arrays.fill(distance, -1);
        
        Queue<Integer> queue = new LinkedList<>();
        distance[start] = 0;
        queue.offer(start);
        
        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            
            for (int neighbor : graph.getNeighbors(vertex)) {
                if (distance[neighbor] == -1) {
                    distance[neighbor] = distance[vertex] + 1;
                    queue.offer(neighbor);
                }
            }
        }
        
        return distance;
    }
    
    // 8️⃣ Dijkstra's Algorithm (weighted graph)
    static class Edge {
        int to, weight;
        Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }
    
    static class WeightedGraph {
        private int V;
        private List<List<Edge>> adj;
        
        WeightedGraph(int V) {
            this.V = V;
            adj = new ArrayList<>();
            for (int i = 0; i < V; i++) {
                adj.add(new ArrayList<>());
            }
        }
        
        void addEdge(int u, int v, int weight) {
            adj.get(u).add(new Edge(v, weight));
        }
        
        List<Edge> getNeighbors(int v) {
            return adj.get(v);
        }
        
        int getVertices() { return V; }
    }
    
    public static int[] dijkstra(WeightedGraph graph, int start) {
        int[] distance = new int[graph.getVertices()];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[start] = 0;
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.offer(new int[]{start, 0});
        
        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int vertex = current[0];
            int dist = current[1];
            
            if (dist > distance[vertex]) continue;
            
            for (Edge edge : graph.getNeighbors(vertex)) {
                int newDist = distance[vertex] + edge.weight;
                if (newDist < distance[edge.to]) {
                    distance[edge.to] = newDist;
                    pq.offer(new int[]{edge.to, newDist});
                }
            }
        }
        
        return distance;
    }
    
    // 9️⃣ Bipartite Graph Check
    public static boolean isBipartite(Graph graph) {
        int[] color = new int[graph.getVertices()];
        Arrays.fill(color, -1);
        
        for (int i = 0; i < graph.getVertices(); i++) {
            if (color[i] == -1) {
                if (!isBipartiteBFS(graph, i, color)) {
                    return false;
                }
            }
        }
        return true;
    }
    
    private static boolean isBipartiteBFS(Graph graph, int start, int[] color) {
        Queue<Integer> queue = new LinkedList<>();
        color[start] = 0;
        queue.offer(start);
        
        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            
            for (int neighbor : graph.getNeighbors(vertex)) {
                if (color[neighbor] == -1) {
                    color[neighbor] = 1 - color[vertex];
                    queue.offer(neighbor);
                } else if (color[neighbor] == color[vertex]) {
                    return false;
                }
            }
        }
        return true;
    }
    
    // 🔟 Strongly Connected Components (Kosaraju's Algorithm)
    public static List<List<Integer>> stronglyConnectedComponents(Graph graph) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] visited = new boolean[graph.getVertices()];
        Stack<Integer> stack = new Stack<>();
        
        // Step 1: Fill stack with vertices in order of finishing times
        for (int i = 0; i < graph.getVertices(); i++) {
            if (!visited[i]) {
                fillOrder(graph, i, visited, stack);
            }
        }
        
        // Step 2: Create transpose graph
        Graph transpose = getTranspose(graph);
        
        // Step 3: Process vertices in reverse order
        Arrays.fill(visited, false);
        while (!stack.isEmpty()) {
            int vertex = stack.pop();
            if (!visited[vertex]) {
                List<Integer> component = new ArrayList<>();
                DFSUtil(transpose, vertex, visited, component);
                result.add(component);
            }
        }
        
        return result;
    }
    
    private static void fillOrder(Graph graph, int vertex, boolean[] visited, Stack<Integer> stack) {
        visited[vertex] = true;
        for (int neighbor : graph.getNeighbors(vertex)) {
            if (!visited[neighbor]) {
                fillOrder(graph, neighbor, visited, stack);
            }
        }
        stack.push(vertex);
    }
    
    private static Graph getTranspose(Graph graph) {
        Graph transpose = new Graph(graph.getVertices());
        for (int i = 0; i < graph.getVertices(); i++) {
            for (int neighbor : graph.getNeighbors(i)) {
                transpose.addEdge(neighbor, i);
            }
        }
        return transpose;
    }
    
    private static void DFSUtil(Graph graph, int vertex, boolean[] visited, List<Integer> component) {
        visited[vertex] = true;
        component.add(vertex);
        for (int neighbor : graph.getNeighbors(vertex)) {
            if (!visited[neighbor]) {
                DFSUtil(graph, neighbor, visited, component);
            }
        }
    }
    
    // 🧭 Demonstration
    public static void main(String[] args) {
        // Create a sample graph
        Graph graph = new Graph(6);
        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 3);
        graph.addEdge(2, 4);
        graph.addEdge(3, 5);
        graph.addEdge(4, 5);
        
        System.out.println("=== Graph Algorithms Demo ===");
        
        // BFS
        BFS(graph, 0);
        
        // DFS
        DFS(graph, 0);
        DFSIterative(graph, 0);
        
        // Cycle detection
        System.out.println("Has cycle (directed): " + hasCycleDirected(graph));
        System.out.println("Has cycle (undirected): " + hasCycleUndirected(graph));
        
        // Topological sort
        List<Integer> topoSort = topologicalSort(graph);
        System.out.println("Topological sort: " + topoSort);
        
        // Shortest path
        int[] distances = shortestPathBFS(graph, 0);
        System.out.println("Shortest distances from 0: " + Arrays.toString(distances));
        
        // Bipartite check
        System.out.println("Is bipartite: " + isBipartite(graph));
        
        // Strongly connected components
        List<List<Integer>> scc = stronglyConnectedComponents(graph);
        System.out.println("Strongly Connected Components: " + scc);
        
        // Weighted graph example
        WeightedGraph wGraph = new WeightedGraph(4);
        wGraph.addEdge(0, 1, 4);
        wGraph.addEdge(0, 2, 1);
        wGraph.addEdge(1, 3, 1);
        wGraph.addEdge(2, 1, 2);
        wGraph.addEdge(2, 3, 5);
        
        int[] dijkstraDistances = dijkstra(wGraph, 0);
        System.out.println("Dijkstra distances from 0: " + Arrays.toString(dijkstraDistances));
    }
}

File: Graph-Algorithms/graph_algorithms.py
"""
🌐 Graph Algorithms - Python Implementation
Author: AI Assistant

Description:
This Python program contains essential graph algorithms frequently asked
in technical interviews at top companies like FAANG, Google, Microsoft, and Amazon.

Each algorithm includes:
- Problem definition
- Clean Python implementation
- Example usage
- Time and Space Complexity
"""

from collections import defaultdict, deque
import heapq
from typing import List, Dict, Set, Tuple, Optional


class Graph:
    """Graph representation using adjacency list"""
    
    def __init__(self, vertices: int):
        self.V = vertices
        self.adj = defaultdict(list)
    
    def add_edge(self, u: int, v: int):
        """Add directed edge from u to v"""
        self.adj[u].append(v)
    
    def add_undirected_edge(self, u: int, v: int):
        """Add undirected edge between u and v"""
        self.adj[u].append(v)
        self.adj[v].append(u)
    
    def get_neighbors(self, v: int) -> List[int]:
        """Get neighbors of vertex v"""
        return self.adj[v]


class WeightedGraph:
    """Weighted graph representation"""
    
    def __init__(self, vertices: int):
        self.V = vertices
        self.adj = defaultdict(list)
    
    def add_edge(self, u: int, v: int, weight: int):
        """Add weighted edge from u to v"""
        self.adj[u].append((v, weight))


def bfs(graph: Graph, start: int) -> List[int]:
    """
    1️⃣ Breadth-First Search (BFS)
    Time: O(V + E), Space: O(V)
    """
    visited = [False] * graph.V
    queue = deque([start])
    visited[start] = True
    result = []
    
    while queue:
        vertex = queue.popleft()
        result.append(vertex)
        
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)
    
    return result


def dfs(graph: Graph, start: int) -> List[int]:
    """
    2️⃣ Depth-First Search (DFS) - Recursive
    Time: O(V + E), Space: O(V)
    """
    visited = [False] * graph.V
    result = []
    
    def dfs_util(vertex):
        visited[vertex] = True
        result.append(vertex)
        
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                dfs_util(neighbor)
    
    dfs_util(start)
    return result


def dfs_iterative(graph: Graph, start: int) -> List[int]:
    """
    3️⃣ DFS - Iterative
    Time: O(V + E), Space: O(V)
    """
    visited = [False] * graph.V
    stack = [start]
    result = []
    
    while stack:
        vertex = stack.pop()
        if not visited[vertex]:
            visited[vertex] = True
            result.append(vertex)
            
            # Add neighbors in reverse order to maintain left-to-right traversal
            for neighbor in reversed(graph.get_neighbors(vertex)):
                if not visited[neighbor]:
                    stack.append(neighbor)
    
    return result


def has_cycle_directed(graph: Graph) -> bool:
    """
    4️⃣ Detect Cycle in Directed Graph (DFS with colors)
    Time: O(V + E), Space: O(V)
    """
    color = [0] * graph.V  # 0: white, 1: gray, 2: black
    
    def has_cycle_dfs(vertex):
        color[vertex] = 1  # gray
        
        for neighbor in graph.get_neighbors(vertex):
            if color[neighbor] == 1:  # back edge
                return True
            if color[neighbor] == 0 and has_cycle_dfs(neighbor):
                return True
        
        color[vertex] = 2  # black
        return False
    
    for i in range(graph.V):
        if color[i] == 0 and has_cycle_dfs(i):
            return True
    
    return False


def has_cycle_undirected(graph: Graph) -> bool:
    """
    5️⃣ Detect Cycle in Undirected Graph
    Time: O(V + E), Space: O(V)
    """
    visited = [False] * graph.V
    
    def has_cycle_dfs(vertex, parent):
        visited[vertex] = True
        
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                if has_cycle_dfs(neighbor, vertex):
                    return True
            elif neighbor != parent:
                return True  # back edge
        
        return False
    
    for i in range(graph.V):
        if not visited[i] and has_cycle_dfs(i, -1):
            return True
    
    return False


def topological_sort(graph: Graph) -> List[int]:
    """
    6️⃣ Topological Sorting (Kahn's Algorithm)
    Time: O(V + E), Space: O(V)
    """
    in_degree = [0] * graph.V
    
    # Calculate in-degrees
    for i in range(graph.V):
        for neighbor in graph.get_neighbors(i):
            in_degree[neighbor] += 1
    
    queue = deque()
    for i in range(graph.V):
        if in_degree[i] == 0:
            queue.append(i)
    
    result = []
    while queue:
        vertex = queue.popleft()
        result.append(vertex)
        
        for neighbor in graph.get_neighbors(vertex):
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)
    
    return result if len(result) == graph.V else []


def shortest_path_bfs(graph: Graph, start: int) -> List[int]:
    """
    7️⃣ Shortest Path - BFS (unweighted graph)
    Time: O(V + E), Space: O(V)
    """
    distance = [-1] * graph.V
    queue = deque([start])
    distance[start] = 0
    
    while queue:
        vertex = queue.popleft()
        
        for neighbor in graph.get_neighbors(vertex):
            if distance[neighbor] == -1:
                distance[neighbor] = distance[vertex] + 1
                queue.append(neighbor)
    
    return distance


def dijkstra(graph: WeightedGraph, start: int) -> List[int]:
    """
    8️⃣ Dijkstra's Algorithm (weighted graph)
    Time: O((V + E) log V), Space: O(V)
    """
    distance = [float('inf')] * graph.V
    distance[start] = 0
    pq = [(0, start)]
    
    while pq:
        dist, vertex = heapq.heappop(pq)
        
        if dist > distance[vertex]:
            continue
        
        for neighbor, weight in graph.adj[vertex]:
            new_dist = distance[vertex] + weight
            if new_dist < distance[neighbor]:
                distance[neighbor] = new_dist
                heapq.heappush(pq, (new_dist, neighbor))
    
    return distance


def is_bipartite(graph: Graph) -> bool:
    """
    9️⃣ Bipartite Graph Check
    Time: O(V + E), Space: O(V)
    """
    color = [-1] * graph.V
    
    def is_bipartite_bfs(start):
        queue = deque([start])
        color[start] = 0
        
        while queue:
            vertex = queue.popleft()
            
            for neighbor in graph.get_neighbors(vertex):
                if color[neighbor] == -1:
                    color[neighbor] = 1 - color[vertex]
                    queue.append(neighbor)
                elif color[neighbor] == color[vertex]:
                    return False
        
        return True
    
    for i in range(graph.V):
        if color[i] == -1:
            if not is_bipartite_bfs(i):
                return False
    
    return True


def strongly_connected_components(graph: Graph) -> List[List[int]]:
    """
    🔟 Strongly Connected Components (Kosaraju's Algorithm)
    Time: O(V + E), Space: O(V)
    """
    def fill_order(vertex, visited, stack):
        visited[vertex] = True
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                fill_order(neighbor, visited, stack)
        stack.append(vertex)
    
    def get_transpose():
        transpose = Graph(graph.V)
        for i in range(graph.V):
            for neighbor in graph.get_neighbors(i):
                transpose.add_edge(neighbor, i)
        return transpose
    
    def dfs_util(vertex, visited, component):
        visited[vertex] = True
        component.append(vertex)
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                dfs_util(neighbor, visited, component)
    
    # Step 1: Fill stack with vertices in order of finishing times
    visited = [False] * graph.V
    stack = []
    for i in range(graph.V):
        if not visited[i]:
            fill_order(i, visited, stack)
    
    # Step 2: Create transpose graph
    transpose = get_transpose()
    
    # Step 3: Process vertices in reverse order
    visited = [False] * graph.V
    result = []
    while stack:
        vertex = stack.pop()
        if not visited[vertex]:
            component = []
            dfs_util(vertex, visited, component)
            result.append(component)
    
    return result


def kruskal_mst(graph: WeightedGraph) -> List[Tuple[int, int, int]]:
    """
    1️⃣1️⃣ Kruskal's Algorithm for MST
    Time: O(E log E), Space: O(V)
    """
    class UnionFind:
        def __init__(self, n):
            self.parent = list(range(n))
            self.rank = [0] * n
        
        def find(self, x):
            if self.parent[x] != x:
                self.parent[x] = self.find(self.parent[x])
            return self.parent[x]
        
        def union(self, x, y):
            px, py = self.find(x), self.find(y)
            if px == py:
                return False
            if self.rank[px] < self.rank[py]:
                px, py = py, px
            self.parent[py] = px
            if self.rank[px] == self.rank[py]:
                self.rank[px] += 1
            return True
    
    # Collect all edges
    edges = []
    for u in range(graph.V):
        for v, weight in graph.adj[u]:
            if u < v:  # Avoid duplicate edges
                edges.append((weight, u, v))
    
    edges.sort()  # Sort by weight
    uf = UnionFind(graph.V)
    mst = []
    
    for weight, u, v in edges:
        if uf.union(u, v):
            mst.append((u, v, weight))
            if len(mst) == graph.V - 1:
                break
    
    return mst


def prim_mst(graph: WeightedGraph) -> List[Tuple[int, int, int]]:
    """
    1️⃣2️⃣ Prim's Algorithm for MST
    Time: O((V + E) log V), Space: O(V)
    """
    mst = []
    visited = [False] * graph.V
    pq = [(0, 0, -1)]  # (weight, vertex, parent)
    
    while pq and len(mst) < graph.V - 1:
        weight, vertex, parent = heapq.heappop(pq)
        
        if visited[vertex]:
            continue
        
        visited[vertex] = True
        if parent != -1:
            mst.append((parent, vertex, weight))
        
        for neighbor, edge_weight in graph.adj[vertex]:
            if not visited[neighbor]:
                heapq.heappush(pq, (edge_weight, neighbor, vertex))
    
    return mst


def articulation_points(graph: Graph) -> Set[int]:
    """
    1️⃣3️⃣ Find Articulation Points (Tarjan's Algorithm)
    Time: O(V + E), Space: O(V)
    """
    ap = set()
    visited = [False] * graph.V
    disc = [0] * graph.V
    low = [0] * graph.V
    time = [0]
    parent = [-1] * graph.V
    
    def dfs(vertex):
        visited[vertex] = True
        disc[vertex] = low[vertex] = time[0]
        time[0] += 1
        children = 0
        
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                children += 1
                parent[neighbor] = vertex
                dfs(neighbor)
                low[vertex] = min(low[vertex], low[neighbor])
                
                # Check if vertex is articulation point
                if parent[vertex] == -1 and children > 1:
                    ap.add(vertex)
                if parent[vertex] != -1 and low[neighbor] >= disc[vertex]:
                    ap.add(vertex)
            elif neighbor != parent[vertex]:
                low[vertex] = min(low[vertex], disc[neighbor])
    
    for i in range(graph.V):
        if not visited[i]:
            dfs(i)
    
    return ap


def bridges(graph: Graph) -> List[Tuple[int, int]]:
    """
    1️⃣4️⃣ Find Bridges (Tarjan's Algorithm)
    Time: O(V + E), Space: O(V)
    """
    bridges_list = []
    visited = [False] * graph.V
    disc = [0] * graph.V
    low = [0] * graph.V
    time = [0]
    parent = [-1] * graph.V
    
    def dfs(vertex):
        visited[vertex] = True
        disc[vertex] = low[vertex] = time[0]
        time[0] += 1
        
        for neighbor in graph.get_neighbors(vertex):
            if not visited[neighbor]:
                parent[neighbor] = vertex
                dfs(neighbor)
                low[vertex] = min(low[vertex], low[neighbor])
                
                # Check if edge is bridge
                if low[neighbor] > disc[vertex]:
                    bridges_list.append((vertex, neighbor))
            elif neighbor != parent[vertex]:
                low[vertex] = min(low[vertex], disc[neighbor])
    
    for i in range(graph.V):
        if not visited[i]:
            dfs(i)
    
    return bridges_list


def hamiltonian_path(graph: Graph) -> Optional[List[int]]:
    """
    1️⃣5️⃣ Hamiltonian Path (Backtracking)
    Time: O(V!), Space: O(V)
    """
    def is_valid_path(path):
        for i in range(len(path) - 1):
            if path[i + 1] not in graph.get_neighbors(path[i]):
                return False
        return True
    
    def hamiltonian_util(path, visited):
        if len(path) == graph.V:
            return path[:]
        
        last_vertex = path[-1]
        for neighbor in graph.get_neighbors(last_vertex):
            if not visited[neighbor]:
                visited[neighbor] = True
                path.append(neighbor)
                
                result = hamiltonian_util(path, visited)
                if result:
                    return result
                
                path.pop()
                visited[neighbor] = False
        
        return None
    
    for start in range(graph.V):
        visited = [False] * graph.V
        visited[start] = True
        result = hamiltonian_util([start], visited)
        if result:
            return result
    
    return None


# 🧭 Demonstration
def main():
    print("=== Graph Algorithms Demo ===")
    
    # Create a sample graph
    graph = Graph(6)
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 3)
    graph.add_edge(2, 4)
    graph.add_edge(3, 5)
    graph.add_edge(4, 5)
    
    # BFS
    print(f"BFS traversal: {bfs(graph, 0)}")
    
    # DFS
    print(f"DFS traversal: {dfs(graph, 0)}")
    print(f"DFS (Iterative): {dfs_iterative(graph, 0)}")
    
    # Cycle detection
    print(f"Has cycle (directed): {has_cycle_directed(graph)}")
    print(f"Has cycle (undirected): {has_cycle_undirected(graph)}")
    
    # Topological sort
    print(f"Topological sort: {topological_sort(graph)}")
    
    # Shortest path
    print(f"Shortest distances from 0: {shortest_path_bfs(graph, 0)}")
    
    # Bipartite check
    print(f"Is bipartite: {is_bipartite(graph)}")
    
    # Strongly connected components
    print(f"Strongly Connected Components: {strongly_connected_components(graph)}")
    
    # Weighted graph example
    w_graph = WeightedGraph(4)
    w_graph.add_edge(0, 1, 4)
    w_graph.add_edge(0, 2, 1)
    w_graph.add_edge(1, 3, 1)
    w_graph.add_edge(2, 1, 2)
    w_graph.add_edge(2, 3, 5)
    
    print(f"Dijkstra distances from 0: {dijkstra(w_graph, 0)}")
    
    # MST algorithms
    print(f"Kruskal MST: {kruskal_mst(w_graph)}")
    print(f"Prim MST: {prim_mst(w_graph)}")
    
    # Articulation points and bridges
    undirected_graph = Graph(5)
    undirected_graph.add_undirected_edge(0, 1)
    undirected_graph.add_undirected_edge(1, 2)
    undirected_graph.add_undirected_edge(2, 3)
    undirected_graph.add_undirected_edge(3, 4)
    undirected_graph.add_undirected_edge(1, 3)
    
    print(f"Articulation points: {articulation_points(undirected_graph)}")
    print(f"Bridges: {bridges(undirected_graph)}")
    
    # Hamiltonian path
    hamiltonian_graph = Graph(4)
    hamiltonian_graph.add_undirected_edge(0, 1)
    hamiltonian_graph.add_undirected_edge(1, 2)
    hamiltonian_graph.add_undirected_edge(2, 3)
    hamiltonian_graph.add_undirected_edge(3, 0)
    hamiltonian_graph.add_undirected_edge(0, 2)
    
    print(f"Hamiltonian path: {hamiltonian_path(hamiltonian_graph)}")


if __name__ == "__main__":
    main()
